cou/apps/base.py

```diff
@@ -912,29 +912,35 @@ def _check_application_target(self, target: OpenStackRelease) -> None:
             )
 
     def _check_mismatched_versions(self) -> None:
-        """Check that there are no mismatched versions on app units.
+        """Check that there are no unexpected mismatched versions on app units.
+
+        For cases where mismatched versions may be expected
+        (eg. nova-compute which is upgraded unit-by-unit),
+        then it will not check.
 
         :raises MismatchedOpenStackVersions: When the units of the app are running
                                              different OpenStack versions.
         """
-        # NOTE (gabrielcocenza) nova-compute is upgraded using paused-single-unit,
-        # so it's possible to have mismatched version in applications units that are
-        # nova-compute or colocated with it.
+        # nova-compute is upgraded one unit at a time,
+        # so it's possible to have mismatched version in
+        # units of applications that are nova-compute or colocated with it.
         if any(
             "nova-compute" in app_charm
             for machine in self.machines.values()
             for app_charm in machine.apps_charms
         ):
             return
 
-        o7k_versions = self.o7k_release_units
-        if len(o7k_versions.keys()) > 1:
-            mismatched_repr = [
-                f"'{openstack_release.codename}': {units}"
-                for openstack_release, units in o7k_versions.items()
-            ]
-
+        if len({self.get_latest_o7k_version(unit) for unit in self.units.values()}) > 1:
+            formatted_results = "\n".join(
+                f"  {unit.name}: {self.get_latest_o7k_version(unit)} "
+                f"(workload: {unit.workload_version})"
+                for unit in sorted(self.units.values(), key=lambda unit: unit.name)
+            )
             raise MismatchedOpenStackVersions(
-                f"Units of application {self.name} are running mismatched OpenStack versions: "
-                f"{', '.join(mismatched_repr)}. This is not currently handled."
+                f"Units of application {self.name} are running mismatched OpenStack releases, "
+                "based on the workload versions as reported by juju status. "
```

Comment:
TODO: juju cli to get this info

Reply:
```
for unit in vault/{0,1,2}; do juju show-unit $unit --format json | jq '."'"${unit}"'"."workload-version"'; done
```

The best I could do is this - anyone know of a more concise way? :) Perhaps this kind of snippet could be put in COU docs for later reference and debugging help?

Reply:
To debug users need to ssh and solve. That could be a simple package update and upgrade or changing sources.

Reply:
Yep, I don't mean to debug though, I mean to manually get the individual units' workload versions. Just with `juju status` it shows a single way so we can't see the version per-unit.

Reply:
COU will do that for you, so I don't see a need to point to a spinet

Reply:
True, I was more curious for my own manual debugging reference.

Reply:
Maybe you would like to check the details implementation of https://ops.readthedocs.io/en/latest/index.html#ops.Unit.set_workload_version ?